text/0000-ternary.md

```diff
@@ -0,0 +1,238 @@
++ Feature name: (fill me in with a unique ident, `my_awesome_feature`)
++ Start date: (fill me in with today's date, YYYY-MM-DD)
++ RFC PR: [iotaledger/bee-rfcs#0000](https://github.com/iotaledger/bee-rfcs/pull/0000)
++ Bee issue: [iotaledger/bee#0000](https://github.com/iotaledger/bee/issues/0000)
+
+# Summary
+
+In this RFC we introduce both the API and implementation of the `bee-ternary`
+crate, a general-purpose ternary manipulation crate written in Rust that may be
+used to handle ternary data in the IOTA ecosystem.
+
+# Motivation
+
+Ternary is a fundamental part of the IOTA ecosystem and is used throughout, even
+for fundamental features like wallet addresses and transaction identification.
+
+Manipulating ternary in an ergonomic manner on top of existing binary-driven
+code can require complexity and, up until now, comes either with a lot of
+boilerplate code, implementations that are difficult to verify as correct, or a
+lack of features.
+
+`bee-ternary` seeks to become the canonical ternary crate in the Rust IOTA
+ecosystem by providing an API that is efficient, ergonomic, and featureful all
+at once.
+
+We hope that `bee-ternary` will allow IOTA developers to more easily write code
+that works with ternary and allow the simplification of existing codebases
+through use of the API.
+
+Broadly, there are 3 main benefits things that `bee-ternary` aims to provide
+over use-specific implementations of ternary manipulation code.
+
+- **Ergonomics**: The API should be trivial to use correctly and should
+naturally guide developers towards correct and efficient solutions.
+
+- **Features**: The API should provide a powerful set of core features that can
+be used together to cover most developer requirements. Examples of such features
+include multiple encoding schemes, binary/ternary conversion, (de)serialization,
+and tryte string de/encoding.
+
+- **Performance**: The API should allow 'acceptably' efficient manipulation of
+ternary data. This is difficult to ruggedly define, and implementation details
+may change later, but broadly we intend to introduce an API that doesn't
+inherently inhibit high performance through poor design choices.
+
+# Detailed design
+
+`bee-ternary` is designed to be extensible and is built on top of a handful of
+powerful abstraction types. Below we list the features of the API and summarise
+a few of the core API objects.
+
+## Features
+
+- Efficient manipulation of ternary buffers (trits and trytes)
+- Multiple encoding schemes
+- Extensible design that allows it to sit on top of existing data structures,
+avoiding unnecessary allocation and copying
+- A plethora of utility functions to allow for easy manipulation of ternary data
+- Zero-cost conversion between trit and tryte formats
```

Comment:
just making sure it's correct what we state here: Is it really 0-cost? If an actual conversion is happening, then I wonder how that can be 0-cost. But if you say that it's just a *reinterpretation* of the underlying bits, then it's fine. But isn't the byte encoding of a trit different from the byte encoding of a tryte, so how can this be 0-cost?

Reply:
It's a reinterpretation in the case of `T3B1`, yes.

Regardless, the canonical usage of 'zero-cost' doesn't mean "requires not computation" - it just means that there's no cost imposed over a manual implementation of the feature (see https://idursun.com/posts/rust_zero_cost_abstractions_in_action/).

Reply:
ah, yea, thanks for the reminder. I still find it a bit confusing (and potentially some readers as well?). Maybe add this link to help readers understand this Rust notion of 0-cost?

Reply:
Will add an explanation